src/lib.rs

```diff
@@ -145,15 +145,220 @@
 extern crate typed_arena;
 
-use doc::Doc::{Annotated, Append, Group, Nest, Newline, Nil, Space, Text};
 use std::borrow::Cow;
 use std::fmt;
+use std::io;
 use std::ops::Deref;
+#[cfg(feature = "termcolor")]
+use termcolor::{ColorSpec, WriteColor};
+
+mod render;
+
+/// The concrete document type. This type is not meant to be used directly. Instead use the static
+/// functions on `Doc` or the methods on an `DocAllocator`.
+///
+/// The `B` parameter is used to abstract over pointers to `Doc`. See `RefDoc` and `BoxDoc` for how
+/// it is used
+#[derive(Clone, Debug, Eq, Ord, PartialEq, PartialOrd)]
+pub enum Doc<'a, B, A = ()> {
+    Nil,
+    Append(B, B),
+    Group(B),
+    Nest(usize, B),
+    Space,
+    Newline,
+    Text(Cow<'a, str>),
+    Annotated(A, B),
+}
+
+impl<'a, A, B> Doc<'a, A, B> {
+    /// An empty document.
+    #[inline]
+    pub fn nil() -> Doc<'a, A, B> {
+        Doc::Nil
+    }
+
+    /// The text `t.to_string()`.
+    ///
+    /// The given text must not contain line breaks.
+    #[inline]
+    pub fn as_string<T: ToString>(t: T) -> Doc<'a, A, B> {
+        Doc::text(t.to_string())
+    }
+
+    /// A single newline.
+    #[inline]
+    pub fn newline() -> Doc<'a, A, B> {
+        Doc::Newline
+    }
+
+    /// The given text, which must not contain line breaks.
+    #[inline]
+    pub fn text<T: Into<Cow<'a, str>>>(data: T) -> Doc<'a, A, B> {
+        let text = data.into();
+        Doc::Text(text)
+    }
+
+    /// A space.
+    #[inline]
+    pub fn space() -> Doc<'a, A, B> {
+        Doc::Space
+    }
+}
+
+impl<'a, A> Doc<'a, BoxDoc<'a, A>, A> {
+    /// Append the given document after this document.
+    #[inline]
+    pub fn append(self, that: Doc<'a, BoxDoc<'a, A>, A>) -> Doc<'a, BoxDoc<'a, A>, A> {
+        DocBuilder(&BOX_ALLOCATOR, self).append(that).into()
+    }
+
+    /// A single document concatenating all the given documents.
+    #[inline]
+    pub fn concat<I>(docs: I) -> Doc<'a, BoxDoc<'a, A>, A>
+    where
+        I: IntoIterator<Item = Doc<'a, BoxDoc<'a, A>, A>>,
+    {
+        docs.into_iter().fold(Doc::nil(), |a, b| a.append(b))
+    }
+
+    /// A single document interspersing the given separator `S` between the given documents.  For
+    /// example, if the documents are `[A, B, C, ..., Z]`, this yields `[A, S, B, S, C, S, ..., S, Z]`.
+    ///
+    /// Compare [the `intersperse` method from the `itertools` crate](https://docs.rs/itertools/0.5.9/itertools/trait.Itertools.html#method.intersperse).
+    #[inline]
+    pub fn intersperse<I, S>(docs: I, separator: S) -> Doc<'a, BoxDoc<'a, A>, A>
+    where
+        I: IntoIterator<Item = Doc<'a, BoxDoc<'a, A>, A>>,
+        S: Into<Doc<'a, BoxDoc<'a, A>, A>> + Clone,
+        A: Clone,
+    {
+        let separator = separator.into();
+        let mut result = Doc::nil();
+        let mut iter = docs.into_iter();
+        if let Some(first) = iter.next() {
+            result = result.append(first);
+        }
+        for doc in iter {
```

Comment:
There is actually a bug here if the iterator is not fused. It should return `result/Doc::nil` if the iterator is empty and not enter the for loop. It existed before this PR though.